File: dcp-solutions/2024-11-30/solution.rb
```ruby
def letterCombinations(digits, mapping = {0 => ['a', 'b', 'c']})
  return [] if digits.empty?

  result = []
  current = ''

  def backtrack(combination)
    unless digits.empty?
      current += combination
      for letter in mapping[digits.first]
        result << letter + combination[1..-1] unless combination == ''
        backtrack(letter)
      end
      current.pop
    else
      result << current
    end
  end

  backtrack('')
  return result
end
```

File: dcp-solutions/2024-11-30/solution.py
```python
def letterCombinations(digits):
    if not digits:
        return []

    phone_mapping = {
        "2": ["a", "b", "c"],
        "3": ["d", "e", "f"],
        "4": ["g", "h", "i"],
        "5": ["j", "k", "l"],
        "6": ["m", "n", "o"],
        "7": ["p", "q", "r", "s"],
        "8": ["t", "u", "v"],
        "9": ["w", "x", "y", "z"]
    }

    def backtrack(combination, next_digits):
        if len(next_digits) == 0:
            output.append(combination)
        else:
            for letter in phone_mapping[next_digits[0]]:
                backtrack(combination + letter, next_digits[1:])

    output = []
    backtrack("", digits)
    return output
```